src/effect/tap_tempo.rs

```diff
@@ -0,0 +1,151 @@
+use crate::{audio::midi::Message, config::NoteOn, effect::Tempo};
+use wmidi::MidiMessage;
+use State::*;
+
+#[derive(Debug)]
+enum State {
+    Start,
+    TappedOnce(u64),
+}
+
+#[derive(Debug)]
+pub struct TapTempo {
+    note_on: NoteOn,
+    state: State,
+}
+
+impl TapTempo {
+    const MAX_DELAY_BETWEEN_TAPS: u64 = 2_000_000; // microseconds
+
+    pub fn new(note_on: NoteOn) -> Self {
+        Self {
+            note_on,
+            state: State::Start,
+        }
+    }
+
+    pub fn handle_messages(&mut self, messages: &[Message]) -> Option<Tempo> {
+        messages
+            .iter()
+            .flat_map(|message| self.handle_message(message))
```

Comment:
In what sense is this a flat-map?

Reply:
In Rust, `flat_map` allows you to map/flatten different kinds of "iterable" things at once. `handle_message` returns an `Option` which is iterable (treating it as a collection which has either 0 or 1 element), and that is being flattened. So we end up with an iterator or `Tempo`, and then take the last one.